File: android/src/main/kotlin/com/selfsign/webview/selfsign_webview_flutter/BridgeRegistry.kt
package com.selfsign.webview.selfsign_webview_flutter

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * Process-wide singleton that fans bridge events from any active WebView
 * (full-screen Activity or embedded PlatformView) to the single Flutter
 * EventChannel sink.
 */
internal object BridgeRegistry : EventChannel.StreamHandler {

    private val mainHandler = Handler(Looper.getMainLooper())
    private var sink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    /**
     * Forward a bridge event to Flutter on the main thread.
     */
    fun send(method: String, payload: String, sourceUrl: String?, viewId: String?) {
        mainHandler.post {
            val map = HashMap<String, Any?>()
            map["method"] = method
            map["payload"] = payload
            map["sourceUrl"] = sourceUrl
            map["viewId"] = viewId
            sink?.success(map)
        }
    }
}

File: ios/Classes/BridgeRegistry.swift
import Flutter
import Foundation

/// Process-wide singleton fan-out from any WKWebView produced by the plugin
/// (full-screen view controller or embedded PlatformView) to a single
/// FlutterEventSink.
final class BridgeRegistry: NSObject, FlutterStreamHandler {
    static let shared = BridgeRegistry()
    private var sink: FlutterEventSink?

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }

    func send(method: String, payload: String, sourceUrl: String?, viewId: String?) {
        let event: [String: Any?] = [
            "method": method,
            "payload": payload,
            "sourceUrl": sourceUrl as Any?,
            "viewId": viewId as Any?
        ]
        if Thread.isMainThread {
            sink?(event)
        } else {
            DispatchQueue.main.async { [weak self] in
                self?.sink?(event)
            }
        }
    }
}
